### PR TITLE
Shapefile: parse in batches

### PR DESCRIPTION
**Pull request by kylebarron**

### CODE REVIEW

File: modules/shapefile/src/lib/parsers/parse-dbf.js [hunk 1]
```diff
@@ -15,13 +15,16 @@ const STATE = {
 class DBFParser {
   constructor({encoding}) {
     this.binaryReader = new BinaryChunkReader();
-    this.textDecoder = new TextDecoder(encoding);
+    this.textDecoder = new TextDecoder(encoding || 'latin1');
```
Comment (kylebarron) [inline modules/shapefile/src/lib/parsers/parse-dbf.js:18]: I found that sometimes when `cpg` files were missing, `null` would be passed through to here, and I learned that in `const {a = 1} = {a: null}`, `null` takes preference over 1.

File: modules/shapefile/src/lib/parsers/parse-shapefile.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      });
+      generator = await zipBatchIterators(shapeIterator, propertyIterator);
+    } else {
+      // Ignore properties
+      generator = shapeIterator;
+    }
+  } catch (error) {
+    // Ignore properties
+    generator = shapeIterator;
+  }
```
Comment (kylebarron) [inline modules/shapefile/src/lib/parsers/parse-shapefile.js:32]: I'm sure there's a better way to fall back than the double-nested `generator = shapeIterator`, and the `if (!propertyIterator)` block below looks a little bad.

File: modules/shapefile/src/lib/parsers/parse-dbf.js [hunk 3]
```diff
@@ -17,11 +17,14 @@ class DBFParser {
...
+    this.result = {
+      data: []
+    };
   }
 
   write(arrayBuffer) {
-    this.binaryReader.write(arrayBuffer);
+    // TODO: fix Node streaming to produce ArrayBuffers, not Buffers
+    this.binaryReader.write(toArrayBuffer(arrayBuffer));
     this.state = parseState(this.state, this.result, this.binaryReader, this.textDecoder);
```
Comment (ibgreen) [inline modules/shapefile/src/lib/parsers/parse-dbf.js:28]: We could just make the binaryReader do the slicing? It is already partially byteOffset aware..

File: modules/shapefile/src/lib/parsers/parse-dbf.js [hunk 4]
```diff
@@ -17,11 +17,14 @@ class DBFParser {
...
     this.state = STATE.START;
-    this.result = {};
+    this.result = {
+      data: []
+    };
   }
 
   write(arrayBuffer) {
-    this.binaryReader.write(arrayBuffer);
+    // TODO: fix Node streaming to produce ArrayBuffers, not Buffers
```
[the remark this answers is not in the document]
Comment (kylebarron) [reply modules/shapefile/src/lib/parsers/parse-dbf.js:26]: Ref #928

File: modules/shapefile/src/lib/parsers/parse-dbf.js (hunk 3 above)
Comment (kylebarron) [reply modules/shapefile/src/lib/parsers/parse-dbf.js:28]: I'm not sure I understand what you're getting at. `parseState` calls `binaryReader.getDataView` internally.

**Pull request merged**